### PR TITLE
Initial Implementation of conditionals 

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

This commit adds an initial implementation for conditionals and depends on #1031.

 
In this implementation, condition evaluations aka ConditionChecks are
backed by TaskRuns. All conditionChecks associated with a `PipelineTask`
have to succeed before the task is executed. If a ConditionCheck fails,
the PipelineTask's associated TaskRun is marked failed i.e. its
`Status.ConditionSucceeded` is False. However, the PipelineRun itself
is not marked as failed.

Future PRs will add full support for passing in parameters, resource contents, and ability to read pipelinerun status from within the conditional container.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) and [TaskRun](../tekton/publish-run.yaml) to build and release this image

# Release Notes

Adds a `Condition` type that can be used to conditionally execute tasks in a pipeline. Docs are  in `./docs/condtions.md` and `./docs/pipelines.md`

### CODE REVIEW

**Comment by bobcatfish** [comment]
Pretty nice coverage boost! :D It looks like the coverage for test/builder/pipeline.go went down which I think means we are missing the test for what you've added to the builder (we do test the tests in this case XD).  Would be nice to cover the additions to test/controller.go but that can be hard 😅

**Comment by m1093782566** [comment]
Nice work!
/cc

**Comment by bobcatfish** [comment]
@dibyom let me know if you want me to review this now - I'm assuming we want to resolve the whole TaskRun question from #1031 before getting too deep into this one.

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution_test.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+			name:  "failed-condition-check",
+			state: failedState,
+			want:  false,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			got := tc.state.IsSuccess()
+			if d := cmp.Diff(got, tc.want); d != "" {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution_test.go]: hm i think it might be simpler to compare `got` and `tc.want` directly since they are both bools?

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go [hunk 2]
```diff
@@ -200,6 +206,15 @@ func (e *ResourceNotFoundError) Error() string {
...
 }
 
+type ConditionNotFoundError struct {
+	Name string
+	Msg  string
+}
+
+func (e *ConditionNotFoundError) Error() string {
+	return fmt.Sprintf("Couldn't retrieve Condition %q: %s", e.Name, e.Msg)
+}
```
[the remark this answers is not in the document]
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: another thought tho: I noticed you moved the condition resolution logic into its own file with it's own tests - instead of having `ResolvePipelineRun` call `resolveConditionChecks`, the reconciler could call `resolveConditionChecks` directly - maybe the problem we're having with returning a specific error could indicate this function has too many responsibilities?

(It would also reduce the number of test cases you need to test at the `ResolvePipelineRun` level!)

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go [hunk 3]
```diff
@@ -261,12 +277,36 @@ func ResolvePipelineRun(
...
 	}
 	return state, nil
 }
 
+// getConditionCheckName should return a unique name for a `ConditionCheck` if one has not already been defined, and the existing one otherwise.
+func getConditionCheckName(taskRunStatus map[string]*v1alpha1.PipelineRunTaskRunStatus, trName, conditionName string) string {
+	trStatus, ok := taskRunStatus[trName]
+	if ok && trStatus.ConditionChecks != nil {
+		for k, v := range trStatus.ConditionChecks {
+			// TODO: We should allow multiple conditions of the same Name type?
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: hm do you mean something like this:

```
conditions:
  - conditionRef: "always-true"
  - conditionRef: "always-true"
```

It's hard to imagine needing this so my inclination is to suggest we don't allow it until it becomes clear we need it?

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		return false
+	}
+	isDone := true
+	for _, rcc := range state {
+		isDone = isDone && !rcc.ConditionCheck.Status.GetCondition(apis.ConditionSucceeded).IsUnknown()
+	}
+	return isDone
+}
+
+func (state TaskConditionCheckState) IsSuccess() bool {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: could a docstring be added for each variable + function exported in this file?

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go [hunk 5]
```diff
@@ -1005,3 +1307,225 @@ func TestReconcilePropagateAnnotations(t *testing.T) {
...
+			tb.TaskRunLabel(pipeline.GroupName+pipeline.PipelineLabelKey, "test-pipeline-run-with-conditions"),
+			tb.TaskRunLabel(pipeline.GroupName+pipeline.PipelineRunLabelKey, "test-pipeline"),
+			tb.TaskRunSpec(tb.TaskRunTaskRef("hello-world")),
+			tb.TaskRunStatus(tb.Condition(apis.Condition{
+				Type:   apis.ConditionSucceeded,
+				Status: corev1.ConditionTrue,
+			}),
+			),
+		),
+		// TODO: add a test builder for conditions/checks
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go]: hee hee my apologies for asking for this in a comment XD 🙏 

If you want to leave it as a TODO, i like using the google type syntax and indicating what issue it'll be fixed as part of, like:

```
// TODO(#12345)
```

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go [hunk 6]
```diff
@@ -419,13 +435,296 @@ func TestUpdateTaskRunsState(t *testing.T) {
 
 }
 
+// TODO: Switch to single test func with diff inputs
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go]: tbh it feels a bit easier to read the way you have it! but i could be convinced otherwise :)

reconciler level tests are a bit hard to follow anyway :S the more we can avoid testing at this level the better!

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 7]
```diff
@@ -365,21 +392,54 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
...
+				}
+				prtrs.Status.SetCondition(&apis.Condition{
+					Type:    apis.ConditionSucceeded,
+					Status:  corev1.ConditionFalse,
+					Reason:  resources.ReasonConditionCheckFailed,
+					Message: fmt.Sprintf("ConditionChecks failed for Task %s in PipelineRun %s", rprt.TaskRunName, pr.Name),
+				})
+			}
+		}
+		pr.Status.TaskRuns[rprt.TaskRunName] = prtrs
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: hm just a thought - if `updateTaskRunsStatus` was instead something like `getTaskRunsStatus` and returned the appropriate status (or statuses?) that should be set, then maybe some of the reconciler level tests could be removed? (e.g. the failing condition case?)

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 8]
```diff
@@ -344,11 +361,21 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
...
+					return xerrors.Errorf("error creating TaskRun called %s for PipelineTask %s from PipelineRun %s: %w", rprt.TaskRunName, rprt.PipelineTask.Name, pr.Name, err)
+				}
+			} else if !rprt.ResolvedConditionChecks.HasStarted() {
+				for _, rcc := range rprt.ResolvedConditionChecks {
+					rcc.ConditionCheck, err = c.makeConditionCheckContainer(c.Logger, rprt, rcc, pr)
+					if err != nil {
+						c.Recorder.Eventf(pr, corev1.EventTypeWarning, "ConditionCheckCreationFailed", "Failed to create TaskRun %q: %v", rcc.ConditionCheckName, err)
+						return xerrors.Errorf("error creating ConditionCheck container called %s for PipelineTask %s from PipelineRun %s: %w", rcc.ConditionCheckName, rprt.PipelineTask.Name, pr.Name, err)
+					}
+				}
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: what do you think, could this chunk of code be a separate function? (outside the reconciler)

File: pkg/apis/pipeline/v1alpha1/condition_types.go [hunk 9]
```diff
@@ -27,6 +27,7 @@ import (
 
 // Check that Task may be validated and defaulted.
 var _ apis.Validatable = (*Condition)(nil)
+//var _ apis.Defaultable = (*Condition)(nil)
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/condition_types.go]: do we want to include this line?

Comment (dibyom) [reply pkg/apis/pipeline/v1alpha1/condition_types.go]: will remove!

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go (hunk 6 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go]: yeah so this comment is from a previous review; I did attempt to make this into a single function but I don't think it was particularly readable - especially the part where we need to add nested maps. I think I'll just remove the TODO

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution_test.go (hunk 1 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution_test.go]: :man_facepalming:

File: test/builder/pipeline.go [hunk 10]
```diff
@@ -399,10 +410,13 @@ func PipelineRunCompletionTime(t time.Time) PipelineRunStatusOp {
...
+// PipelineRunTaskRunsStatus sets the status of TaskRun to the PipelineRunStatus.
+func PipelineRunTaskRunsStatus(taskRunName string, status *v1alpha1.PipelineRunTaskRunStatus) PipelineRunStatusOp {
 	return func(s *v1alpha1.PipelineRunStatus) {
-		s.TaskRuns = taskRuns
+		if s.TaskRuns == nil {
+			s.TaskRuns = make(map[string]*v1alpha1.PipelineRunTaskRunStatus)
+		}
+		s.TaskRuns[taskRunName] = status
 	}
 }
```
[the remark this answers is not in the document]
Comment (dibyom) [reply test/builder/pipeline.go]: So, looks like there is already a `Condition` type in the builder package: https://github.com/tektoncd/pipeline/blob/master/test/builder/task.go#L315

I'm thinking we can rename it to `StatusCondition` or `TaskRunStatusCondition` and call this one Condition? (Will probably do that in a different PR since we'd be renaming a bunch of functions)

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution_test.go [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+	{
+		ConditionCheckName: "foo",
+		Condition:          c,
+		ConditionCheck: &v1alpha1.ConditionCheck{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "running-condition-check",
+			},
+		},
+	},
+}
```
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution_test.go]: Fixed!

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go (hunk 3 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: We might need it for something like two `stringEquals` conditions together. Though its definitely more of an advanced use case we can tackle in https://github.com/tektoncd/pipeline/issues/1022

File: test/builder/pipeline.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dibyom) [reply test/builder/pipeline.go]: Opened #1127

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go (hunk 4 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: done

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+	if !state.IsComplete() {
+		return false
+	}
+	isSuccess := true
+	for _, rcc := range state {
+		isSuccess = isSuccess && rcc.ConditionCheck.Status.GetCondition(apis.ConditionSucceeded).IsTrue()
+	}
+	return isSuccess
+}
```
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: Done. Though we still need the functions here since they operate on a slice of `ResolvedConditionChecks`

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 7 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: Done. Also switched the 3 status tests with condition checks to one with multiple inputs.

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: The main reason for adding the `ConditionNotFound` error type (besides tests) is to return an error when a PipelineRun fails due to a missing Condition: https://github.com/tektoncd/pipeline/pull/1093/files#diff-225ed18b17b573cba99d592dea1da1a0R290

I was following the pattern there where we return different error types for Resources or Tasks not being found.

[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: `ResolvePipelineRun` definitely does a lot (e.g. resolving tasks, taskruns resources, and now conditions and checks). I'm open to ideas on simplifying/modularizing this. At the moment though, we need this in order to build the full `PipelineRunState` in one call which is then used by the reconciler.

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 8 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: Simplified this slightly (reduced one level of indentation). This block requires a bunch of the reconciler field (logger, recoder, the clientset). In the create Taskrun function, we use the clientset for both creating as well as updating, so it'll be hard to get it out of the reconciler. I'll play around with this some more, but what do you think about refactoring this in a follow-up?

File: test/builder/pipeline.go [hunk 13]
```diff
@@ -399,10 +410,13 @@ func PipelineRunCompletionTime(t time.Time) PipelineRunStatusOp {
...
-// PipelineRunTaskRunsStatus sets the TaskRuns of the PipelineRunStatus.
-func PipelineRunTaskRunsStatus(taskRuns map[string]*v1alpha1.PipelineRunTaskRunStatus) PipelineRunStatusOp {
+// PipelineRunTaskRunsStatus sets the status of TaskRun to the PipelineRunStatus.
+func PipelineRunTaskRunsStatus(taskRunName string, status *v1alpha1.PipelineRunTaskRunStatus) PipelineRunStatusOp {
 	return func(s *v1alpha1.PipelineRunStatus) {
-		s.TaskRuns = taskRuns
+		if s.TaskRuns == nil {
+			s.TaskRuns = make(map[string]*v1alpha1.PipelineRunTaskRunStatus)
+		}
+		s.TaskRuns[taskRunName] = status
```
Comment (bobcatfish) [inline test/builder/pipeline.go]: i think this change might not be covered in the test? (judging by the coverage metric decreasing)

File: docs/conditions.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
+# Conditions
+
+This document defines `Conditions` and their capabilities.
+
+*NOTE*: This feature is currently a WIP.
```
Comment (bobcatfish) [inline docs/conditions.md]: maybe link here to the related issue(s)?

File: docs/conditions.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+---
+
+- [Syntax](#syntax)
+  - [Check](#check)
+  - [Parameters](#parameters)
```
Comment (bobcatfish) [inline docs/conditions.md]: i think there isn't a "parameters" heading yet?

File: docs/conditions.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    the spec must include:
+    - [`check`](#check) - Specifies a container that you want to run for evaluating the condition 
+
+[kubernetes-overview]:
+  https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields
+
+### Check
+
+The `check` field is required. You define a single check to define the body of a `Condition`. The 
+check must specify a container image that adheres to the container contract. The container image
```
Comment (bobcatfish) [inline docs/conditions.md]: "container contract" could link to https://github.com/tektoncd/pipeline/blob/master/docs/container-contract.md

File: docs/pipelines.md [hunk 17]
```diff
@@ -282,6 +284,28 @@ In this example, the task "build-the-image" will be executed and if the first
 run fails a second one would triggered. But, if that fails no more would
 triggered: a max of two executions.
 
+
+#### conditions
+
+Sometimes you will need to run tasks only when some conditions are true. The `conditions` field 
+allows you to list a series of references to [`Conditions`](#conditions) that are run before the task
```
Comment (bobcatfish) [inline docs/pipelines.md]: i think you might want `(conditions.md)` instead of `(#conditions)` here? (to link to the conditions doc vs linking back to the section we're already in)

File: docs/pipelines.md [hunk 18]
```diff
@@ -282,6 +284,28 @@ In this example, the task "build-the-image" will be executed and if the first
...
+is run. If all of the conditions evaluate to true, the task is run. If any of the conditions are false,
+the Task is not run. Its status.ConditionSucceeded is set to False with the reason set to  `ConditionCheckFailed`.
+However, unlike regular task failures, condition failures do not automatically fail the entire pipeline 
+run -- other tasks in the pipeline are still run.
```
Comment (bobcatfish) [inline docs/pipelines.md]: might be good to state what happens here to subsequent Tasks that are `runAfter` or `from` this Task (I would assume that if a Task is using a resource `from` this Task, it couldn't possibly run? not as clear what the `runAfter` behaviour should be...)

Comment (bobcatfish) [reply docs/pipelines.md]: and i noticed that that case isn't present in the examples or the end to end tests - would probably be good to include somewhere! (or maybe in a follow up PR? but even in that case maybe we could document what our intended behaviour is)

File: pkg/apis/pipeline/register.go [hunk 19]
```diff
@@ -18,10 +18,11 @@ package pipeline
...
-	PipelineLabelKey     = "/pipeline"
-	PipelineRunLabelKey  = "/pipelineRun"
-	PipelineTaskLabelKey = "/pipelineTask"
+	GroupName                    = "tekton.dev"
+	TaskLabelKey                 = "/task"
+	TaskRunLabelKey              = "/taskRun"
+	PipelineLabelKey             = "/pipeline"
+	PipelineRunLabelKey          = "/pipelineRun"
+	PipelineTaskLabelKey         = "/pipelineTask"
+	PipelineRunConditionCheckKey = "/pipelineConditionCheck"
```
Comment (bobcatfish) [inline pkg/apis/pipeline/register.go]: might be worth including docstrings for these (since they get exported)

File: pkg/apis/pipeline/v1alpha1/condition_types_test.go [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+			Type:   apis.ConditionSucceeded,
+			Status: corev1.ConditionTrue,
+		},
+	)))
+
+	cc := v1alpha1.ConditionCheck(*tr)
+	if !cc.IsSuccessful() {
+		t.Fatal("Expected conditionCheck status to be done")
+	}
+}
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/condition_types_test.go]: nice :D

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 8 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: >  I'll play around with this some more, but what do you think about refactoring this in a follow-up?

Sounds good to me :D Thanks for looking into it :)

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+package resources
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+const (
+	unnamedCheckNamePrefix = "condition-check-"
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: im not quite clear what "unnamed" means here 🤔  (maybe a comment?)

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go]: > ResolvePipelineRun definitely does a lot (e.g. resolving tasks, taskruns resources, and now conditions and checks). I'm open to ideas on simplifying/modularizing this. At the moment though, we need this in order to build the full PipelineRunState in one call which is then used by the reconciler.

kk maybe we can refactor it in follow up PRs! For me the most appealing option is to have multiple `Resolve` functions, each resolving a different type of thing, each being called by the Reconciler - then the Reconciler will know what kind of error message to set cuz it will know what kind of thing its trying to reconcile. Not super important tho!

File: test/pipelinerun_test.go [hunk 22]
```diff
@@ -113,6 +114,32 @@ func TestPipelineRun(t *testing.T) {
 		// 1 from PipelineRun and 1 from Tasks defined in pipelinerun
 		expectedNumberOfEvents: 2,
 		pipelineRunFunc:        getHelloWorldPipelineRun,
+	}, {
+		name: "pipeline succeeds when task skipped due to failed condition",
```
Comment (bobcatfish) [inline test/pipelinerun_test.go]: thanks for adding this! i think i derped a little bit when I suggested this be an end to end test vs. an example: i though the pipelinerun would end up failing (therefore wouldnt work as an example) but i forgot that a failing condition actually doesnt affect the success of the PipelineRun 🤦‍♀  anyway tho i think either is fine :D

File: test/builder/pipeline.go (hunk 13 above)
Comment (dibyom) [reply test/builder/pipeline.go]: done!

File: docs/conditions.md (hunk 16 above)
Comment (dibyom) [reply docs/conditions.md]: done

File: docs/pipelines.md (hunk 17 above)
Comment (dibyom) [reply docs/pipelines.md]: done

File: pkg/apis/pipeline/register.go (hunk 19 above)
Comment (dibyom) [reply pkg/apis/pipeline/register.go]: done!

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go (hunk 21 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: added

File: docs/conditions.md (hunk 14 above)
Comment (dibyom) [reply docs/conditions.md]: done

File: docs/pipelines.md (hunk 18 above)
Comment (dibyom) [reply docs/pipelines.md]: clarified that this is only for tasks that are not dependent. 
I think for `runAfter` we should be explicit and let the user select what to do. I have #1023 open for that.

File: test/pipelinerun_test.go (hunk 22 above)
Comment (dibyom) [reply test/pipelinerun_test.go]: Well, this led me down a rabbit hole which led to #1130 which I think was what finally fixed the issue where creating and using a Condition immediately would always fail with a NotFound error. So, definitely worth it :angel:

File: docs/pipelines.md (hunk 18 above)
Comment (bobcatfish) [reply docs/pipelines.md]: kk sounds good!

File: docs/conditions.md [hunk 23]
```diff
@@ -0,0 +1,50 @@
+# Conditions
+
+This document defines `Conditions` and their capabilities.
+
+*NOTE*: This feature is currently a WIP being tracked in [#1137](https://github.com/tektoncd/pipeline/issues/1137)
```
Comment (bobcatfish) [inline docs/conditions.md]: 🎉

File: pkg/apis/pipeline/register.go [hunk 24]
```diff
@@ -16,12 +16,25 @@ limitations under the License.
...
+	// PipelineLabelKey is used as the label identifier for a Pipeline
+	PipelineLabelKey = "/pipeline"
+
+	// PipelineRunLabelKey is used as the label identifier for a PipelineRun
+	PipelineRunLabelKey = "/pipelineRun"
+
+	// PipelineRunLabelKey is used as the label identifier for a PipelineTask
 	PipelineTaskLabelKey = "/pipelineTask"
+
+	// ConditionCheck is used as the label identifier for a ConditionCheck
```
Comment (bobcatfish) [inline pkg/apis/pipeline/register.go]: thanks for adding these!!
